pkg/controller/quota/quota.go

```
@@ -0,0 +1,165 @@
+
+	// Create a condition setter for AppInstanceConditionQuotaAllocated, which blocks the appInstance from being deployed
+	// until quota has been allocated.
+	status := condition.Setter(appInstance, resp, apiv1.AppInstanceConditionQuotaAllocated)
+
+	// Don't do anything if quota isn't enabled for this project
+	if enforced, err := isEnforced(req, appInstance.Namespace); !enforced {
+		if err != nil {
+			status.Error(err)
+			return err
+		}
+		status.Success()
+		return nil
+	}
+
+	// Attempt to get the quotaRequest for this appInstance. It should exist with the name and namespace of the
+	// appInstance being processed.
+	quotaRequest := &adminv1.QuotaRequestInstance{}
+	err := req.Client.Get(req.Ctx, router.Key(appInstance.Namespace, appInstance.Name), quotaRequest)
+	if err != nil && !errors.IsNotFound(err) {
+		return err
+	}
+
+	// Determine how to proceed depending on if the quotaRequest exists and what it has written to its status
+	if quotaRequest.Status.FailedResources != nil {
+		resp.DisablePrune()
+		status.Error(fmt.Errorf("failed to provision the following resources: %v", quotaRequest.Status.FailedResources.ToString()))
+	} else if quotaRequest.Status.Condition(adminv1.QuotaRequestCondition).Error {
+		status.Error(fmt.Errorf("error occurred while trying to allocate quota: %w", err))
+	} else if err != nil || !equality.Semantic.DeepEqual(quotaRequest.Spec.Resources, quotaRequest.Status.AllocatedResources) {
+		status.Unknown("waiting for quota allocation")
+	} else if quotaRequest.Status.Condition(adminv1.QuotaRequestCondition).Success {
+		status.Success()
+	}
+
+	return nil
+}
+
+// EnsureQuotaRequest ensures that the quota request exists and is up to date.
+func EnsureQuotaRequest(req router.Request, resp router.Response) error {
+	appInstance := req.Object.(*apiv1.AppInstance)
+
+	// Don't do anything if quota isn't enabled for this project
+	if enforced, err := isEnforced(req, appInstance.Namespace); !enforced {
```

looks like the err is not being checked